### ISSUE TITLE
Questions about custom objective function in XGBoost

### ISSUE THREAD

**Issue by liupei101** [ytsaig/notebooks#1]
Hello, ytsaig! 

I have no idea about how to contact you, so I opened issues here !

Since I am doing implementation of multiclass custom objective, so I carefully read you comments on the issue [#2113 Missing example of multiclass custom objective](https://github.com/dmlc/xgboost/issues/2113) in XGBoost, and also viewed the code you published on [gist](https://gist.github.com/ytsaig/a596feec53d2a024ac69f5ae5a83d8f7).

I have question that: 
Do we need to do softmax transform on `preds` before computing `grad` and `hess` in your code of the function `softmaxobj(preds, dtrain)` ? and does the corresponding code need to be updated ?

The Concern is from the reference [custom_objective.py](https://github.com/dmlc/xgboost/blob/master/demo/guide-python/custom_objective.py) updated by the author of Tianqi Chen, details of which are as follows:
```
# user defined evaluation function, return a pair metric_name, result
# NOTE: when you do customized loss function, the default prediction value is margin
# this may make builtin evaluation metric not function properly
# for example, we are doing logistic loss, the prediction is score before logistic transformation
# the builtin evaluation error assumes input is after logistic transformation
# Take this in mind when you use the customization, and maybe you need write customized evaluation function
```

Look forward to your reply ! Sorry for disturbing!
Thx !
